rs/orchestrator/src/utils.rs

```diff
@@ -35,3 +39,140 @@ fn endpoint_to_url(protocol: &str, http: &ConnectionEndpoint, log: &ReplicaLogge
         }
     }
 }
+
+/// Enum for tracking the node assignment states that can happen on the
+/// network.
+#[derive(PartialEq, Eq, Debug)]
+pub(crate) enum NodeAssignationState {
+    /// The node is unassigned in the registry and the node thinks it is unassigned.
+    Unassigned,
+    /// The node is unassigned in the registry but still participates in consensus.
+    Leaving { subnet_id: SubnetId },
+    /// The node is assigned in the registry but the node still hasn't synced the state.
+    Joining { subnet_id: SubnetId },
+    /// The node is assigned in the registry and the node is participating in conseusns.
+    Assigned { subnet_id: SubnetId },
+    /// The node is assigned to subnet A in the registry but the node is participating
+    /// in consensus of subnet B.
+    Moving { from: SubnetId, to: SubnetId },
+    /// Initial state
+    Unknown,
+}
+
+pub(crate) fn maybe_notify_control_flow_change(
+    previous_state: NodeAssignationState,
+    reported_flow: OrchestratorControlFlow,
+    node_id: NodeId,
+    maybe_subnet_from_registry: Option<SubnetId>,
+    log: &ReplicaLogger,
+) -> NodeAssignationState {
+    // Based on the registry view of the node and self view figure out the next state
+    let new_state = match (reported_flow, maybe_subnet_from_registry) {
+        (OrchestratorControlFlow::Unassigned, None) => NodeAssignationState::Unassigned,
+        (OrchestratorControlFlow::Unassigned, Some(s)) => {
+            NodeAssignationState::Joining { subnet_id: s }
+        }
+        (OrchestratorControlFlow::Assigned(s), None) => {
+            NodeAssignationState::Leaving { subnet_id: s }
+        }
+        (OrchestratorControlFlow::Assigned(s_node), Some(s_reg)) => {
+            if s_node == s_reg {
+                NodeAssignationState::Assigned { subnet_id: s_node }
+            } else {
+                NodeAssignationState::Moving {
+                    from: s_node,
+                    to: s_reg,
+                }
+            }
+        }
+        (OrchestratorControlFlow::Stop, _) => {
+            // Not important as the node is gracefully being shutdown in order to upgrade.
+            return previous_state;
+        }
+    };
+
+    // Based on the transition from previous_state -> new_state
+    // figure out what to do.
+    match (&previous_state, &new_state) {
+        (NodeAssignationState::Unknown, s) if s != &NodeAssignationState::Unknown => {
+            // Maybe log the inital thing...
+        }
+        (NodeAssignationState::Unassigned, NodeAssignationState::Joining { subnet_id }) => {
+            // Maybe log node joining subnet...
+            info!(log, "Detected node joining a subnet {subnet_id}");
+        }
```

Comment:
FYI: We can't detect this at the moment, or at least I couldn't figure it out for now. It would be nice to be able to display this so we know how long did it take for the node to sync the state from other nodes and started participating in consensus (If it is already possible sorry, just point me to it and I will incorporate it here).

ATM we get this from the logs for joining the subnet:

<img width="1531" height="180" alt="Image" src="https://github.com/user-attachments/assets/a13ca187-56eb-4123-996f-b3c171cc5398" />